staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/maplist.go

```diff
@@ -0,0 +1,194 @@
+)
+
+// mapList provides a "lookup by key" operation for lists (arrays) with x-kubernetes-list-type=map.
+type mapList interface {
+	// get returns the unique element having identical values, for all
+	// x-kubernetes-list-map-keys, to the provided object. If no such unique element exists, or
+	// if the provided object isn't itself a valid mapList element, get returns nil.
+	get(interface{}) interface{}
+}
+
+type keyStrategy interface {
+	// CompositeKeyFor returns a composite key for the provided object, if possible, and a
+	// boolean that indicates whether or not a key could be generated for the provided object.
+	CompositeKeyFor(map[string]interface{}) (interface{}, bool)
+}
+
+// singleKeyStrategy is a cheaper strategy for associative lists that have exactly one key.
+type singleKeyStrategy struct {
+	key     string
+	defawlt interface{} // default is a keyword
+}
+
+// CompositeKeyFor directly returns the value of the single key (or its default value, if absent) to
+// use as a composite key.
+func (ks *singleKeyStrategy) CompositeKeyFor(obj map[string]interface{}) (interface{}, bool) {
+	v, ok := obj[ks.key]
+	if !ok {
+		v = ks.defawlt // substitute default value
+	}
+
+	switch v.(type) {
+	case bool:
+	case float64:
+	case int64:
+	case string:
+	default:
+		return nil, false // non-scalar
+	}
+
+	return v, true
+}
+
+// hashKeyStrategy computes a hash of all key values.
+type hashKeyStrategy struct {
+	sts    *schema.Structural
+	hasher maphash.Hash
+}
+
+// CompositeKeyFor returns a hash computed from the values (or default values, if absent) of all
+// keys.
+func (ks *hashKeyStrategy) CompositeKeyFor(obj map[string]interface{}) (interface{}, bool) {
+	const keyDelimiter = "\x00" // 0 byte should never appear in the hash input except as delimiter
+
+	ks.hasher.Reset()
+	for _, key := range ks.sts.XListMapKeys {
+		v, ok := obj[key]
+		if !ok {
+			v = ks.sts.Properties[key].Default.Object
+		}
+
+		switch v.(type) {
+		case bool:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%t", v)
+		case float64:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%f", v)
+		case int64:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%d", v)
+		case string:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%q", v)
+		default:
+			return nil, false // values must be scalars
+		}
+	}
+	return ks.hasher.Sum64(), true
+}
+
+// emptyMapList is a mapList containing no elements.
+type emptyMapList struct{}
+
+func (emptyMapList) get(interface{}) interface{} {
+	return nil
+}
+
+type mapListImpl struct {
+	sts      *schema.Structural
+	ks       keyStrategy
+	elements map[interface{}][]interface{} // composite key -> bucket
+}
+
+func (a *mapListImpl) get(obj interface{}) interface{} {
+	mobj, ok := obj.(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	key, ok := a.ks.CompositeKeyFor(mobj)
+	if !ok {
+		return nil
+	}
+
+	// Scan bucket to handle key collisions and duplicate key sets:
+	var match interface{}
+	for _, element := range a.elements[key] {
```

Comment:
Would it be a premature optimization to special case where there is only one element in the hash bucket and return it?

Reply:
That sounds likely to be faster since collisions should be rare. I think that introduces a remote risk, though: the composite key of the argument to `get` may collide with, but have different key values from, the composite key of an existing map entry (even if no other map entries have the same composite key).

Reply:
Probably another case we should defer to when we have time to benchmark and optimize.